README.md
gt-download-upload-wrapper
===================

#Description

This tool is used to monitor and make gt-download and upload ore robust.

To be use primarily with the PanCancer project.

Contained are two modules one for each uploading and downloading


#Usage

##Sample Upload

    GNOS::Upload->run_upload($sub_path, $key, $max_attempts, $timeout_minutes, $ktimeout);

##Sample Download

    GNOS::Download->run_download($class, $pem, $url, $file, $max_attempts, $timeout_minutes, $max_children, $rate_limit_mbytes, $ktimeout);

We are including the GNOS::Download library by using the -I flag with perl that allows you to specify a path to find the repos. Alterantively you can add the file by adding the line "use lib <lib-path>;" or add the path to the @INC Environment variable.

We are also packaging the taged versions of the repo in OICR's articatory for use with Maven projects. This can be very useful for use with SeqWare workflows. Ask any of the developers on this project to find out where you can get the necessary XML to get the artifact with Maven.

Feel free to use these modules in your own scripts.

Currently they are intended to be used in both the BWA and VCF SeqWare Workflows for the PanCancer Project


CHANGELOG.md
# 2.0.13

* upload: fixed a problem where uploads not at 100% are reported as being successful
* download: added an option where a minimal transfer rate is maintaned otherwise the download is retried

# 2.0.12

* added $max_children, $rate_limit_mbytes, $ktimeout to download. These correspond to the following gtdownload params:

    --max-children
    --rate-limit
    -k

# 2.0.11

* added `$ktimeout` to upload, this is the `-k` parameter from gtupload
* added `$max_children`, `$rate_limit_mbytes`, `$ktimeout` to download. These correspond to the following gtdownload params:
    * --max-children <gtdownload_default>
    * --rate-limit <gtdownload_default>
    * -k <minutes_of_inactivity_to_abort_recommend_less_than_timeout_if_you_want_this_to_be_used>
* Previously the values were hard coded to: 4, 200, and 60 respectively for the above, if you want to match the old behavior you must specify these values otherwise the defaults for gtdownload will be used, see [here](https://cghub.ucsc.edu/docs/user/CGHubUserGuide.pdf).


lib/GNOS/Upload.pm
package GNOS::Upload;

use warnings;
use strict;

use feature qw(say);
use autodie;
use Carp qw( croak );

use constant {
    MILLISECONDS_IN_AN_HOUR => 3600000,
};

#############################################################################################
# DESCRIPTION                                                                               #
#############################################################################################
#  This module is wraps the gtupload script and retries the uploads if it freezes up.       #
#############################################################################################
# USAGE: run_upload($sub_path, $key, $max_attempts, $timeout_minutes);                      #
#        Where the command is the full gtuplaod command                                     #
#############################################################################################

sub run_upload {
    my ($class, $sub_path, $key, $max_attempts, $timeout_minutes, $ktimeout) = @_;

    $max_attempts //= 30;
    $timeout_minutes //= 60;
    my $ktimeout_txt = "";
    if ($ktimeout > 0) {
      $ktimeout_txt = " -k $ktimeout ";
    }

    my $timeout_milliseconds = ($timeout_minutes / 60) * MILLISECONDS_IN_AN_HOUR;
    say "TIMEOUT: $timeout_minutes minutes ( $timeout_milliseconds milliseconds )";

    my ($log_filepath, $time_stamp, $pid, $read_output, $command);
    my $attempt = 0;
    do {
        my @now = localtime();
        $time_stamp = sprintf("%04d-%02d-%02d-%02d-%02d-%02d",
                                 $now[5]+1900, $now[4]+1, $now[3],
                                 $now[2],      $now[1],   $now[0]);

        # BUG: Adam, you are already in this directory, $sub_path does not exist!
        #$log_filepath = "$sub_path/gtupload-$time_stamp.log";
        $log_filepath = "gtupload-$time_stamp.log";

        say "STARTING UPLOAD WITH LOG FILE $log_filepath ATTEMPT ".++$attempt." OUT OF $max_attempts";

        my $upload_cmd = "cd $sub_path; gtupload -l $log_filepath $ktimeout_txt -v -c $key -u ./manifest.xml";

        say "UPLOAD COMMAND: $upload_cmd\n";

        `$upload_cmd </dev/null >/dev/null 2>&1 &`;

        sleep 30;

        $read_output = read_output($sub_path, $log_filepath, $timeout_milliseconds);
        if ($read_output == 1 ) {
            say "KILLING PROCESS";
            `sudo pkill -f gtupload -l $log_filepath`;  # need to have log_filepath included so that it doesn't kill other gt_upload process
        }

    } while ( ($attempt < $max_attempts) and ( $read_output ) );

    return 0 if ( ($read_output == 0) and (say "UPLOADED FILE AFTER $attempt ATTEMPTS") );

    say "FAILED TO UPLOAD FILE AFTER $attempt ATTEMPTS";
    return 1;

}

sub read_output {
    my ($sub_path, $log_filepath, $timeout) = @_;

    my $start_time = time;
    my $time_last_uploading = 0;
    my $last_reported_percent = 0;

    my (@lines, $process, $output);

    say "DIR: ".`pwd`." LOG: ".`ls -lth $sub_path/$log_filepath`."\n";

    my ($uploaded, $percent, $rate);
    while(  $output = `tail -n 20 $sub_path/$log_filepath`  ) {
        sleep(10);
        ($uploaded , $percent, $rate) = (0,0,0);

        # Gets last occurance of the progress line in the 20 lines from the tail command
        @lines = split "\n", $output;
        foreach my $line (@lines) {
            if (my @captured = $line =~ m/Status:\s+(\d+.\d+|\d+| )\s+[M|G]B\suploaded\s*\((\d+.\d+|\d+| )%\s*complete\)\s*current\s*rate:\s*(\d+.\d+|\d+| )\s*[M|k]B\/s/g ) {
                ($uploaded, $percent, $rate) = @captured;
            }
        }

        $percent = $last_reported_percent unless( defined $percent);

        $process = `ps aux | grep 'gtupload -l $log_filepath'`; #check to see if the command that we are monitoring has completed.
        return 0 unless ($process =~ m/manifest/); # This checks to see if the gtupload process is still running. Does not say if completed correctly

        if ($percent > $last_reported_percent) {
            $time_last_uploading = time;
            say "  UPLOADING TIME: $time_last_uploading";
            say "  REPORTED PERCENT UPLOADED - LAST: $last_reported_percent CURRENT: $percent";
        }
        elsif ((($time_last_uploading != 0) and ( (time - $time_last_uploading) > $timeout) )
                 or ( ($percent == 0) and ( (time - $start_time) > (3 * $timeout)) )) {
                # This should trigger if gtupload stops being able to upload for a certain amount of time
                #     or if it never starts uploading
            say "BASED ON OUTPUT UPLOAD IS NEEDING TO BE RESTARTED";
            return 1;
        }

        $last_reported_percent = $percent;
    }

    return 1;
}

1;


lib/GNOS/Download.pm
package GNOS::Download;

use warnings;
use strict;

use feature qw(say);
use autodie;
use Carp qw( croak );

use constant {
    MILLISECONDS_IN_AN_HOUR => 3600000,
};

#############################################################################################
# DESCRIPTION                                                                               #
#############################################################################################
#  This module is wraps the gtdownload script and retries the downloads if it freezes up.   #
#############################################################################################
# USAGE: run_download($class, $pem, $url, $file, $max_attempts, $timeout_minutes);          #
#        Where the command is the full gtdownlaod command                                   #
#############################################################################################

# TODO: add min rate

sub run_download {
    my ($class, $pem, $url, $file, $max_attempts, $timeout_minutes, $max_children, $rate_limit_mbytes, $ktimeout) = @_;

    $max_attempts //= 30;
    $timeout_minutes //= 60;
    my $max_children_txt = "";
    if ($max_children > 0) {
      $max_children_txt = "--max-children $max_children";
    }
    my $rate_limit_mbytes_txt = "";
    if ($rate_limit_mbytes > 0) {
      $rate_limit_mbytes_txt = "--rate-limit $rate_limit_mbytes";
    }
    my $ktimeout_txt = "";
    if ($ktimeout > 0) {
      $ktimeout_txt = "-k $ktimeout";
    }

    my $timeout_milliseconds = ($timeout_minutes / 60) * MILLISECONDS_IN_AN_HOUR;
    say "TIMEOUT: $timeout_minutes minutes ( $timeout_milliseconds milliseconds )";

    my ($log_filepath, $time_stamp, $pid, $random_int);
    my $attempt = 0;
    do {
        my @now = localtime();
        $time_stamp = sprintf("%04d-%02d-%02d-%02d-%02d-%02d",
                                 $now[5]+1900, $now[4]+1, $now[3],
                                 $now[2],      $now[1],   $now[0]);

        $random_int = int(rand(1000));
        $log_filepath = "gtdownload-$time_stamp-$random_int.log";
        say "STARTING DOWNLOAD WITH LOG FILE $log_filepath ATTEMPT ".++$attempt." OUT OF $max_attempts";

        `gtdownload -l $log_filepath $max_children_txt $rate_limit_mbytes_txt -c $pem -v $url $ktimeout_txt </dev/null >/dev/null 2>&1 &`;

        sleep 10; # to give gtdownload a chance to make the log files.

        if ( read_output($log_filepath, $timeout_milliseconds) ) {
            say "KILLING PROCESS";
            `sudo pkill -f 'gtdownload -l $log_filepath'`;
        }
        sleep 10; # to make sure that the file has been created.
    } while ( ($attempt < $max_attempts) and ( not (-e $file) ) );

    return 0 if ( (-e $file) and (say "DOWNLOADED FILE $file AFTER $attempt ATTEMPTS") );

    say "FAILED TO DOWNLOAD FILE: $file AFTER $attempt ATTEMPTS";
    return 1;
}

sub read_output {
    my ($log_filepath, $timeout) = @_;

# needs to be param
my $min_rate = 0;

    my $start_time = time;
    my $time_last_downloading = 0;
    my $last_reported_percent = 0;
    my $last_reported_rate = 0;

    my ($size, $percent, $rate, $rate_units);
    $rate_units = "M";
    my (@lines, $output, $process);
    sleep (20); # to wait for gtdownload to create the log file

    while( $output = `tail -n 20 $log_filepath` ) {
        sleep 10;

        ($size , $percent, $rate, $rate_units) = (0,0,0,"M");

        # Gets last occurance of the progress line in the 20 lines from the tail command
        @lines = split "\n", $output;
        foreach my $line (@lines) {
            if (my @captured = $line =~  m/Status:\s*(\d+.\d+|\d+|\s*)\s*[M|G]B\s*downloaded\s*\((\d+.\d+|\d+|\s)%\s*complete\)\s*current rate:\s+(\d+.\d+|\d+| )\s+([M|K|G])B\/s/g) {
                ($size, $percent, $rate, $rate_units) = @captured;
                if ($rate_units eq "K") { $rate = $rate / 1024; }
                if ($rate_units eq "G") { $rate = $rate * 1024; }
            }
        }

        $percent = $last_reported_percent unless( defined $percent);
        $rate = $last_reported_rate unless( defined $rate);

        my $md5sum = ($output =~ m/Download resumed, validating checksums for existing data/g)? 1: 0;

        $process = `ps aux | grep 'gtdownload -l $log_filepath'`;
        return 0 if ($process !~ m/children/ && $percent >= 100); # This checks to see if the gtdownload process is still running. Does not say if completed correctly

        #return 0 if ($percent > 100); # this is an edge case where for some reason the percentage continues increasing beyond 100%

        # LEFT OFF WITH: need to check the rate here... if < threshold for > retries then kill the job
        # need to properly parse the... need to make this check optional

        if ( ( $percent > $last_reported_percent && $rate > $min_rate ) || $md5sum) {  # Checks to see if the download is making progress.
            $time_last_downloading = time;
            say "UPDATING LAST DOWNLOAD TIME: $time_last_downloading";
            say "  REPORTED PERCENT DOWNLOADED - LAST: $last_reported_percent CURRENT: $percent" if ($percent > $last_reported_percent);
            say "  IS MD5Sum State: $md5sum" if ($md5sum);
        }
        elsif ((($time_last_downloading != 0) and ( (time - $time_last_downloading) > $timeout) )
                 or ( ($percent == 0) and ( (time - $start_time) > (3 * $timeout)) )) { # this check is to see the download has not been working for too long
                # This should trigger if gtdownload stops being able to download for a certain amount of time
                #     or if it never starts downloading - giving time for the md5 check
            say "BASED ON OUTPUT DOWNLOAD IS NEEDING TO BE RESTARTED";
            return 1;
        }

        $last_reported_percent = $percent;
    }

    return 1;
}

1;


gtdownload/gnos_download_file.pl
#!/usr/bin/env perl

use warnings;
use strict;

use feature qw(say);

use autodie;

use Getopt::Long;

use GNOS::Download;

# Example Command:
# cd /mnt/seqware-oozie/23f5253a-f33b-11e3-8add-8589c49f5d8e
# perl /mnt/home/seqware/git/genomic_tools/gnos_tools/launch_and_monitor_gnos.pl --command 'gtdownload --max-children 4 --rate-limit 200 -c /home/seqware/provisioned-bundles/Workflow_Bundle_BWA_2.6.3_SeqWare_1.1.0-alpha.5/Workflow_Bundle_BWA/2.6.3/scripts/gnostest.pem -v -d https://gtrepo-dkfz.annailabs.com/cghub/data/analysis/download/23f5253a-f33b-11e3-8add-8589c49f5d8e' --file-grep 23f5253a-f33b-11e3-8add-8589c49f5d8e --search-path . --md5-retries 120 --retries 30
### setup / INSTALL
# mkdir /mnt/seqware-oozie/23f5253a-f33b-11e3-8add-8589c49f5d8e
# sudo apt-get install libcommon-sense-perl
# PURPOSE:
# the program takes a command (use single quotes to encapsulate it in bash) and
# a to check, the existance of which indidates the gtdownload finished running.
# It also takes a retries count, timeout for no data being written in minutes, and a
# cooldown (sleep) time in minutes for checking if the output exists. It then executes the command in a thread and
# watches the files every sleep time. For every timeout period where there is no
# change in the output file size then the retries count is
# decremented. By default, the retries are 30 and timeout 60 minutes so the code
# could retry a download that times out a total of 30 hours before giving up.
# If the retries are exhausted the thread is killed, the
# thread is recreated and started, and the process starts over.

# where file is the GNOS id of the file

my ($pem, $url, $file);
my $cooldown_min = 1;
my $timeout_min = 60;
my $retries = 30;

GetOptions (
"pem=s" => \$pem,
"url=s" => \$url,
"file=s" => \$file,
"retries=i" => \$retries,
"sleep-min=i" => \$cooldown_min,
"timeout-min=i" => \$timeout_min,
);

say "FILE: $file";

# will return 0 on success, not 0 on failure
my $ret_val = GNOS::Download->run_download($pem, $url, $file, $retries, $cooldown_min, $timeout_min);
exit ($ret_val);
